bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+    // with the same addr from each other
+    pub unique_id: u64,
+}
+
+// Established connection management code running in its own task
+struct EstablishedConn {
+    peer_id: Baseboard,
+    unique_id: u64,
+    write_sock: OwnedWriteHalf,
+    read_sock: OwnedReadHalf,
+    main_tx: mpsc::Sender<ConnToMainMsg>,
+    rx: mpsc::Receiver<MainToConnMsg>,
+    log: Logger,
+    read_buf: Vec<u8>,
+    total_read: usize,
+
+    // Used for managing inactivity timeouts for the coonnection
+    last_received_msg: Instant,
+
+    // Keep a queue to write serialized messages into We limit the queue
+    // size, and if it gets exceeded it means the peer at the other
+    // end isn't pulling data out fast enough. This should be basically
+    // impossible to hit given the size and rate of message exchange
+    // between peers. We go ahead and close the connection if the queue
+    // fills.
+    write_queue: VecDeque<Vec<u8>>,
+
+    // The current serialized message being written if there is one
+    current_write: Cursor<Vec<u8>>,
+}
+
+impl EstablishedConn {
+    fn new(
+        peer_id: Baseboard,
+        unique_id: u64,
+        write_sock: OwnedWriteHalf,
+        read_sock: OwnedReadHalf,
+        main_tx: mpsc::Sender<ConnToMainMsg>,
+        rx: mpsc::Receiver<MainToConnMsg>,
+        log: Logger,
+    ) -> EstablishedConn {
+        EstablishedConn {
+            peer_id,
+            unique_id,
+            write_sock,
+            read_sock,
+            main_tx,
+            rx,
+            log,
+            read_buf: vec![0u8; CONN_BUF_SIZE],
+            total_read: 0,
+            last_received_msg: Instant::now(),
+            write_queue: VecDeque::with_capacity(MSG_WRITE_QUEUE_CAPACITY),
+            current_write: Cursor::new(Vec::new()),
+        }
+    }
+
+    // Run the main loop of the connection
+    //
+    // The task can only return a `ConnErr`, otherwise it runs forever
+    async fn run(&mut self) -> ConnErr {
+        let mut interval = interval(PING_INTERVAL);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        loop {
+            if !self.current_write.has_remaining()
+                && !self.write_queue.is_empty()
+            {
+                self.current_write =
+                    Cursor::new(self.write_queue.pop_front().unwrap());
+            }
+
+            let res = tokio::select! {
+                _ = interval.tick() => {
+                    self.ping().await
+                }
+                Some(msg) = self.rx.recv() => {
+                    self.on_msg_from_main(msg).await
+                }
+                res = self.read_sock.read(&mut self.read_buf[self.total_read..]) => {
+                    self.read(res).await
+                }
+                res = self.write_sock.write_buf(&mut self.current_write),
+                   if self.current_write.has_remaining() => {
+                        self.check_write_result(res).await
+                }
```

Comment:
Indentation is a little funky here, maybe
```suggestion
                    }
```

Reply:
I don't think that's correct. The `if` is a precondition on the tokio expression, not standalone. I actually got confused by this yesterday also when re-reading. It's unfortunate.